packages/cli/lib/validators/marketplaceValidators/theme/ModuleValidator.js

```diff
@@ -0,0 +1,129 @@
+const fs = require('fs');
+const BaseValidator = require('../BaseValidator');
+const { isModuleFolderChild } = require('@hubspot/cli-lib/modules');
+
+const MODULE_LIMIT = 50;
+
+class ModuleValidator extends BaseValidator {
+  constructor(options) {
+    super(options);
+
+    this.errors = {
+      LIMIT_EXCEEDED: {
+        key: 'limitExceeded',
+        getCopy: ({ limit, total }) =>
+          `Cannot exceed ${limit} modules in your theme (found ${total})`,
+      },
+      MISSING_META_JSON: {
+        key: 'missingMetaJSON',
+        getCopy: ({ modulePath }) =>
+          `Missing a meta.json file for ${modulePath}`,
+      },
+      INVALID_META_JSON: {
+        key: 'invalidMetaJSON',
+        getCopy: ({ modulePath }) =>
+          `Invalid json in meta.json file for ${modulePath}`,
+      },
+      MISSING_LABEL: {
+        key: 'missingLabel',
+        getCopy: ({ modulePath }) =>
+          `The meta.json file is missing a "label" field for ${modulePath}`,
+      },
+      MISSING_ICON: {
+        key: 'missingIcon',
+        getCopy: ({ modulePath }) =>
+          `The meta.json file is missing an "icon" field for ${modulePath}`,
+      },
+    };
+  }
+
+  getUniqueModulesFromFiles(files) {
+    const uniqueModules = {};
+
+    files.forEach(file => {
+      if (isModuleFolderChild({ isLocal: true, path: file })) {
+        // Get unique module path by removing the file name
+        const lastSlashIndex = file.lastIndexOf('/');
+        const modulePath = file.slice(0, lastSlashIndex);
+        if (!uniqueModules[modulePath]) {
+          uniqueModules[modulePath] = {};
+        }
+        const fileName = file.slice(lastSlashIndex + 1, file.length);
+        uniqueModules[modulePath][fileName] = file;
+      }
+    });
+    return uniqueModules;
+  }
+
+  // Validates:
+  // - Theme does not have more than MODULE_LIMIT modules
+  // - Each module folder contains a meta.json file
+  // - Each module meta.json file contains valid json
+  // - Each module meta.json file has a "label" field
+  // - Each module meta.json file has an "icon" field
+  validate(absoluteThemePath, files) {
+    let validationErrors = [];
+    const uniqueModules = this.getUniqueModulesFromFiles(files);
+    const numModules = Object.keys(uniqueModules).length;
+
+    if (numModules > MODULE_LIMIT) {
+      validationErrors.push(
+        this.getError(this.errors.LIMIT_EXCEEDED, {
+          limit: MODULE_LIMIT,
+          total: numModules,
+        })
+      );
+    }
+
+    Object.keys(uniqueModules).forEach(modulePath => {
+      const metaJSONFile = uniqueModules[modulePath]['meta.json'];
+      const relativePath = modulePath.replace(`${absoluteThemePath}/`, '');
```

Comment:
If we stored the uniqueModules as an array of the relative filepaths, this could be a bit easier.

Reply:
Let me know if I'm misunderstanding what you're suggesting, but I think what I have in place now is simpler than that.

I posted an example of the current shape of `uniqueModules` [here](https://github.com/HubSpot/hubspot-cli/pull/504#discussion_r636160025).

Right now I'm using this object for two things. The first thing I need is to find the total number of modules, which I can do by counting the number of keys in the `uniqueModules` object (which should each correspond to a module in the theme). The second thing I need to do is find the path to the `meta.json` file for each module. Storing the files as an array would mean that I'd need to parse the array of files for each module to find the `meta.json` file. As it is now, I can look up the `meta.json` file by doing something like this `uniqueModules[<module path>]['meta.json']`